lib/closure_tree/hierarchy_maintenance.rb

```diff
@@ -131,9 +130,46 @@ def cleanup!
           lonely_childs = hierarchy_class.joins(arel_join).where(alias_table[primary_key].eq(nil))
           ids = lonely_childs.pluck(foreign_key)
 
-          hierarchy_class.where(hierarchy_table[foreign_key].in(ids)).delete_all
+          q = hierarchy_class.where(hierarchy_table[foreign_key].in(ids))
+          q.delete_all
         end
       end
     end
+
+    private
+
+    def build_hierarchy(parent_nodes, children_nodes)
+      all_nodes = [ClosureTree::Hierarchy::NodeRelation.new(id, send(_ct.parent_column_sym))] + parent_nodes + children_nodes
+      tree = ClosureTree::Hierarchy::TreeBuilder.new(_ct, all_nodes).build
+      ClosureTree::Hierarchy::HierarchyBuilder.new(tree, parent_nodes, all_nodes).build
+    end
+
+    def parent_nodes(node)
+      return [] if node.parent_reference.nil?
+
+      results = _ct.connection.execute <<-SQL.squish
+        SELECT #{_ct.primary_key_column.name}, #{_ct.quoted_parent_column_name} 
+        FROM #{_ct.quoted_table_name} 
+        WHERE #{_ct.primary_key_column.name} = #{_ct.quoted_value(node.parent_reference)}
+      SQL
+
+      return [] if results.size == 0
+
+      parent = ClosureTree::Hierarchy::NodeRelation.from_query_result(_ct, results.first)
+      [parent] + parent_nodes(parent)
+    end
+
+    def children_nodes(nodes)
+      results = _ct.connection.execute <<-SQL.squish
+        SELECT #{_ct.primary_key_column.name}, #{_ct.quoted_parent_column_name}
+        FROM #{_ct.quoted_table_name}
+        WHERE #{_ct.quoted_parent_column_name} IN (#{nodes.map { |node| _ct.quoted_value(node.reference) }.join(',')})
+      SQL
+
+      return [] if results.size == 0
+
+      children = results.map { |result| ClosureTree::Hierarchy::NodeRelation.from_query_result(_ct, result) }
```

Comment:
maybe we can extract here the result data depending on the type of the result here (ie MySQL returns a MySQL result type with an array for each result